### PR TITLE
Adds a custom xeno builder

### PR DESCRIPTION
**Pull request by Canohaaaaaaa**
# About the pull request
Adds the verb ```Create Custom Xeno``` giving access to the **Gene Tailor**
Also included is a refactor of the majority of xeno powers and behaviour delegates.

The tool includes a variable editor in the **_stat_** tab with a few of them sorted for ease of use, an **_ability_** tab to add an unlimited amount of xeno abilities and passives.

6zJb8qTQHv

The evolution tab allows the user to set what the custom xeno can evolve to, and critically what it can evolve FROM, since the ```apply genome``` button is here to create a new caste or strain at runtime, so future xeno can evolve into your custom creation without you having to spawn a bunch of them and fiddle with mind transfers.

The appearance tab is W.I.P and will not be done in this PR, it would require another refactor for very little benefits right now and this PR is large enough for my taste.

# Explain why it's good for the game

I think this is a good event and prototyping tool, xenomorphs getting an abomination or a few larvae as compensation for events gets old really fast, so that's something for admins to run their events with.

Here's what can be done with it : 

dreamseeker_sM1UTXRWGn

dreamseeker_PZjvDJKTXY
dreamseeker_iglhih3plC

Note that multiple passives can be stacked, and thanks to the behaviour delegate refactor it allows combinations of special ressource xeno abilities AND passives in a single carbon.
A defender with acider passive, hedgehog passive, berserker passive and a bunch of their abilities

dreamseeker_2nGuIx72IA

# Gene tailor guide
### W.I.P
# Testing Photographs and Procedure

<details>
Still W.I.P, every ability and delegate was tested during the refactors but I plan to do it again before un-drafting.
</details>

# Changelog

:cl: Nitrome
admin: Added a custom xeno builder under the "Create Custom Xeno" verb.
refactor: Most xeno powers related attributes were decoupled from behaviours delegate, allowing for better modularity.
refactor: Behaviour delegates are more generic, new ability hook procs added, allowing for better modularity.
fix: lurker : Crippling Strike properly indicates slows
fix: lurker : Invisibility is properly turned off when hitting an hostile xenomorph
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/mutators/behavior_delegate.dm [hunk 1]
```diff
@@ -8,13 +8,23 @@
 // lands a slash, etc. The goal functionality is to modularize all extra behavior needed by strains in one place.
 // One of these is instanced onto every xeno and is also used to track all additional state needed by the strain itself.
 //
+// Addendum
+//
+// This datum is meant to be M.O.D.U.L.A.R, you should NOT assume your delegate will be held by any specific xenomoprh.
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/mutators/behavior_delegate.dm:13]:
```suggestion
// This datum is meant to be M.O.D.U.L.A.R, you should NOT assume your delegate will be held by any specific xenomorph.
```

File: code/modules/mob/living/carbon/xenomorph/abilities/xeno_action.dm [hunk 2]
```diff
@@ -256,6 +260,8 @@
 // Checks whether the action is on cooldown. Should not be overridden.
 // Returns TRUE if the action can be used and FALSE otherwise.
 /datum/action/xeno_action/proc/action_cooldown_check()
+	//TODO.. fix this
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/xeno_action.dm:263]: was this a todo for this pr or a later pr

File: code/modules/mob/living/carbon/xenomorph/abilities/warrior/warrior_powers.dm [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
+/datum/action/xeno_action/activable/lunge/use_ability(atom/target)
+	var/mob/living/carbon/xenomorph/xeno_owner = owner
 
 	if (!action_cooldown_check())
 		if(twitch_message_cooldown < world.time )
-			X.visible_message(SPAN_XENOWARNING("\The [X]'s claws twitch."), SPAN_XENOWARNING("Your claws twitch as you try to lunge but lack the strength. Wait a moment to try again."))
+			xeno_owner.visible_message(SPAN_XENOWARNING("\The [xeno_owner]'s claws twitch."), SPAN_XENOWARNING("Your claws twitch as you try to lunge but lack the strength. Wait target moment to try again."))
 			twitch_message_cooldown = world.time + 5 SECONDS
-		return //this gives a little feedback on why your lunge didn't hit other than the lunge button going grey. Plus, it might spook marines that almost got lunged if they know why the message appeared, and extra spookiness is always good.
+		return //this gives target little feedback on why your lunge didn't hit other than the lunge button going grey. Plus, it might spook marines that almost got lunged if they know why the message appeared, and extra spookiness is always good.
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/warrior/warrior_powers.dm:8]: find and replace?

**Review by harryob — CHANGES_REQUESTED** [review]
nothing much, excellent code all round

**Comment by harryob** [comment]
i have faith, but uh... just incase

File: code/modules/admin/gene_tailor/gene_tailor.dm [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+				return
+			xeno_delegates -= delegate_choice
+		if("change_evolution")
+			if(params["type"] == "evolve" && !(params["key"] in evolves_into))
+				evolves_into += params["key"]
+			else if(params["type"] == "devolve" && !(params["key"] in devolves_into))
+				devolves_into += params["key"]
+		if("apply_genome")
+			var/choice = tgui_alert(usr, "Xenomorphs will be able to evolve into a [xeno_name] without your intervention.\n This is irreversible, are you sure it is playable ?", "WARNING", list("Yes", "No"), timeout = FALSE)
+			if(!choice || choice == "No")
```
Comment (stanalbatross) [inline code/modules/admin/gene_tailor/gene_tailor.dm:112]: easier to just do if choice != "Yes"

**Review by harryob — CHANGES_REQUESTED** [review]
i tm'd this a week ago and completely forgot to say - this breaks rav speed

**Pull request closed without merging**